src/playlist-loader.js

```diff
@@ -424,17 +424,46 @@
       return 0;
     }
 
-    for (i = 0; i < this.media_.segments.length; i++) {
-      time -= Playlist.duration(this.media_,
-                                this.media_.mediaSequence + i,
-                                this.media_.mediaSequence + i + 1,
-                                false);
-
-      // HLS version 3 and lower round segment durations to the
-      // nearest decimal integer. When the correct media index is
-      // ambiguous, prefer the higher one.
-      if (time <= 0) {
-        return i;
+    // 1) Walk backward until we find the latest segment with timeline
+    // information that is earlier than `time`
+    targetDuration = this.media_.targetDuration || 10;
+    i = this.media_.segments.length;
+    while (i--) {
+      segment = this.media_.segments[i];
+      if (segment.end !== undefined && segment.end <= time) {
+        time -= segment.end;
+        break;
+      }
+      if (segment.start !== undefined && segment.start < time) {
+
+        if (segment.end !== undefined && segment.end > time) {
+          // we've found the target segment exactly
+          return i;
+        }
+
+        time -= segment.start;
+        time -= segment.duration || targetDuration;
+        break;
+      }
+    }
+    i++;
+
+    // 2) Walk forward, testing each segment to see if `time` falls within it
+    for (j = i; j < this.media_.segments.length; j++) {
+      segment = this.media_.segments[j];
+      time -= segment.duration || targetDuration;
+
+      if (time < 0) {
+        return j;
+      }
+
+      // 2a) If we discover a segment that has timeline information
+      // before finding the result segment, the playlist information
+      // must have been inaccurate. Start a binary search for the
+      // segment which contains `time`. If the guess turns out to be
+      // incorrect, we'll have more info to work with next time.
+      if (segment.start !== undefined || segment.end !== undefined) {
+        return Math.floor((j - i) * 0.5);
```

Comment:
I feel like we could be smarter than a binary search here but I am trying to find a way to explain it..

Basically we know the last point we have accurate information on (the `start_point`) Starting with the lemma that "every `segment.end` is the next segment's `start`" - we either know the `start` of `i--` (when we find a known start) or the `start` of `i` (when we find a known end). (It's `i--` or `i` because you have already incremented `i` after the loop.)

Likewise, we just found a segment we know about so now we know: either the `start` of segment j (the case where `segment.start` is not `undefined`) or the `start` of segment j++ (in the case where `segment.end` is not `undefined`).

With that information we know exactly the number of segments between two known points (either `i+j`, `i+j+1` or `i+j+2`) and we know exactly the range - from one known start point to the other known start point.

With that we can calculate `i + Math.floor(numSegments * (range / time))` to get a much better idea of which segment we are likely in. Again, the same convergence is as with the binary search will occur if we happen to be wrong but we are much more likely to be right the first time and, as a consequence, likely to converge on the correct answer more quickly.